lib/membrane_audio_interleaver.ex

```diff
@@ -0,0 +1,283 @@
+defmodule Membrane.AudioInterleaver do
+  @moduledoc """
+  Element responsible for interleaving several mono audio streams into single interleaved stream.
+  All input streams should be in the same raw audio format, defined by 'caps' option.
+
+  Channels are interleaved in order given in 'order' option - currently required, no default available.
+
+  Each input pad should be identified with your custom id (using 'via_in(Pad.ref(:input, your_example_id)' )
+  """
+
+  use Membrane.Filter
+  use Bunch
+
+  alias Membrane.AudioMixer.DoInterleave
+  alias Membrane.Buffer
+  alias Membrane.Caps.Audio.Raw, as: Caps
+
+  require Membrane.Logger
+
+  def_options caps: [
+                type: :struct,
+                spec: Caps.t(),
+                description: """
+                The value defines a raw audio format of pads connected to the
+                element. It should be the same for all the pads.
+                """,
+                default: nil
+              ],
+              frames_per_buffer: [
+                type: :integer,
+                spec: pos_integer(),
+                description: """
+                Assumed number of raw audio frames in each buffer.
+                Used when converting demand from buffers into bytes.
+                """,
+                default: 2048
+              ],
+              order: [
+                type: :list,
+                spec: list(integer),
+                description: """
+                Order in which channels should be interleaved
+                """,
+                default: nil
+              ]
+
+  def_output_pad :output,
+    mode: :pull,
+    availability: :always,
+    caps: Caps
+
+  def_input_pad :input,
+    mode: :pull,
+    availability: :on_request,
+    demand_unit: :bytes,
+    caps: Caps
+
+  @impl true
+  def handle_init(%__MODULE__{} = options) do
+    state =
+      options
+      |> Map.from_struct()
+      |> Map.put(:pads, %{})
+      |> Map.put(:finished, false)
+      |> Map.put(:channels, length(options.order))
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_pad_added(pad, _context, state) do
+    state =
+      Bunch.Access.put_in(
+        state,
+        [:pads, pad],
+        %{queue: <<>>, stream_ended: false}
+      )
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_pad_removed(pad, _context, state) do
+    state = Bunch.Access.delete_in(state, [:pads, pad])
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_prepared_to_playing(_context, %{caps: %Caps{} = caps, channels: channels} = state) do
+    {{:ok, caps: {:output, %Caps{caps | channels: channels}}}, state}
+  end
+
+  @impl true
+  def handle_prepared_to_playing(_context, %{caps: nil} = state) do
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_demand(:output, size, :bytes, _context, %{channels: channels} = state) do
+    do_handle_demand(div(size, channels), state)
```

Comment:
I am not sure if `div` is enough here. If `div(size, channels) < size / channels`, demand will not be fulfilled. Additionally, `div(size, channels)` is not aware of sample size, so the element may receive only a fragment of the last sample and it will be unable to perform interleaving on that last sample (so less to send).

Reply:
fixed

Reply:
So after discussion with Bartek it can stay as it is

Reply:
after discussion, I went back to div option